cryptodoc/src/08_signatures.rst

```diff
@@ -656,3 +658,317 @@ operations.
+
+.. _signatures/sphincsplus/hypertree:
+
+Hypertree
+^^^^^^^^^
+
+The XMSS hypertree signature creation according to Algorithm 12 of [SPX-R3]_
+(:math:`\mathtt{ht\_sign}`) is implemented by the method ``ht_sign``. Beginning
+at the hypertree's leaves, the hypertree is built up using subsecutive calls of
+``xmss_sign_and_pkgen`` (see :ref:`SPHINCS+ XMSS <signatures/sphincsplus/XMSS>`)
+with each call signing the root of the previous XMSS tree or the hypertree
+signature's message for the first call. As described in :ref:`SPHINCS+ XMSS
+<signatures/sphincsplus/XMSS>`, this also creates the XMSS root node used in the
+next iteration. The leaf indices selected to sign the hypertree signature's
+message or roots are computed according to the specification.
+
+The hypertree verification, Algorithm 13  of [SPX-R3]_
+(:math:`\mathtt{ht\_verify}`), is performed in ``ht_verify``. By calling
+``compute_root``, it reconstructs the roots from bottom to top using the
+concatenated XMSS signatures. For verification, the final root, which is the
+root of the hypertree, is compared with :math:`\mathbf{PK}.\mathsf{root}`.
+
+.. _signatures/sphincsplus/sphincsplus:
+
+SPHINCS\ :sup:`+`
+^^^^^^^^^^^^^^^^^
+
+All the above components are combined to constitute Botan's SPHINCS\ :sup:`+`
+component used for creating or verifying SPHINCS\ :sup:`+` signatures.
+
+Signature Creation
+~~~~~~~~~~~~~~~~~~
+
+A SPHINCS\ :sup:`+` signature is created in the following manner, following
+Algorithm 20 of [SPX-R3]_:
+
+.. admonition:: SPHINCS+ Signature Creation
+
+   **Input:**
+
+   -  ``m``: message to be signed
+   -  ``SK``: SPHINCS\ :sup:`+` secret key, ``SK = {secret_seed, prf, public_seed, sphincs_root}``
+
+   **Output:**
+
+   -  ``sig``:  SPHINCS\ :sup:`+` signature
+
+   **Steps:**
+
+   1. ``opt_rand`` is set to ``SK.public_seed``. If the scheme is randomized, ``opt_rand`` is instead set to a freshly generated random byte vector.
+   2. ``msg_random_s = PRF_msg(m, SK.prf, opt_rand)`` and append ``msg_random_s`` to ``sig``.
+   3. ``mhash || tree_idx || leaf_idx = H_msg(msg_random_s, SK.sphincs_root, m)``.
+   4. Set type of ``fors_addr`` to FORS tree, its tree to ``tree_idx``, and its keypair address to ``leaf_idx``.
+   5. ``fors_sig, fors_root = fors_sign_and_pkgen(mhash, SK.secret_seed, fors_addr)`` and append ``fors_sig`` to ``sig``.
+   6. ``ht_sig = ht_sign(fors_root, SK.secret_seed, tree_idx, leaf_idx)`` and append ``ht_sig`` to ``sig``.
+
+   **Notes:**
+
+   - ``SK.public_seed`` is omitted as an input because the hash functions are already instantiated with a corresponding member variable.
+
+Signature Validation
+~~~~~~~~~~~~~~~~~~~~
+
+A SPHINCS\ :sup:`+` signature is verified in the following manner, following
+Algorithm 21 of [SPX-R3]_:
+
+.. admonition:: SPHINCS+ Signature Validation
+
+   **Input:**
+
+   -  ``m``: message to be validated
+   -  ``sig``: signature to be validated
```

Comment:
Thanks! Also, for consistency I'll name it `SPHINCS+ signature`. The relationship "of `m`" is stated in the output description:
`-  ``true``, if the signature for message ``m`` is valid. ``false`` otherwise`
IMO this is sensible, because the validation method can be called with valid and invalid signatures.